---
Title: "AtrialMap"

Outline: This is the demo code for AtrialMap, including "Filtering and doublet removal", "Data integration and cell-cycle regression", "Single-cell clustering", "Cell type identification and annotation", "Functional analysis of cell populations", "Cell–cell communication", "Gene score calculation" and "Cell trajectory analysis".

A brief explanation: Many different datasets were used in this work, like Atrial_Nuclei_Donor.rdata, Atrial_Total_Cells_Donor.rdata, Atrial_CD45_cells_Donor.rdata, and so on. To keep things simple and avoid confusion, "Object" or "test" was used as global name for these datasets. Duplicate codes or loops were deleted or simplified as soon as possible.
---

Loading libraries
```{r}
# if (!require("BiocManager", quietly = TRUE))
    # install.packages("BiocManager")
#BiocManager::install(Packages)
library(Seurat)
library(cluster) 
library(factoextra)
library(ComplexHeatmap)
library(circlize)
library(magrittr)
library(tidyverse)
library(ggplot2)
library(viridis)
library(clusterProfiler)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(enrichplot)
library(DOSE)
library(DoubletFinder)
library(biomaRt)
library(CellChat)
library(ggplot2)
library(ggalluvial)
library(svglite)
library(nichenetr)
library(tidyverse)
library(monocle)
```

Preparing data
```{r}
counts <- Read10X(data.dir = "~/raw_data")
Object <- CreateSeuratObject(counts = counts,project = "project name", min.features=200)
Object <- RenameCells(object = Object, add.cell.id = "project name")
# for human 
Object[["percent.mt"]] <- PercentageFeatureSet(object = Object, pattern = "^MT-")
# for mouse
Object[["percent.mt"]] <- PercentageFeatureSet(object = Object, pattern = "^mt-")
```

Filtering
```{r}
# data from different patients or treatment groups (sham and TAC group) was processed separately
VlnPlot(object = Object , features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
# cells expressed less than 200 genes (low quality) or more than 4,000 genes (potential doublets) were excluded for further analysis
Object_QC<-subset(x = Object, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & 
                    nCount_RNA < 10^(mean(log10(Object$nCount_RNA))+2*sd(log10(Object$nCount_RNA))) & 
                    nCount_RNA > 10^(mean(log10(Object$nCount_RNA))-2*sd(log10(Object$nCount_RNA))) & 
                    percent.mt < 20)
dim(Object_QC) 
Object_QC[["source"]] <- "source"
Object_QC[["condition"]] <- "disease condition"
```

Checking ERCC and MT- 
```{r}
ERCC<-grep("^ERCC",rownames(Object_QC))
dim(Object_QC)  
Object_QC_no_ERCC <- Object_QC[!rownames(Object_QC) %in% rownames(Object_QC)[ERCC],]
dim(Object_QC_no_ERCC) 
grep("^ERCC",rownames(Object_QC_no_ERCC))

MT<-grep("^MT-",rownames(Object_QC_no_ERCC))
Object_QC_no_ERCC_MT<-Object_QC_no_ERCC[!rownames(Object_QC_no_ERCC) %in% rownames(Object_QC_no_ERCC)[MT],]
dim(Object_QC_no_ERCC_MT)
grep("^MT-",rownames(Object_QC_no_ERCC_MT))
```

Removing doublet
```{r}
test_doublet <- SCTransform(Object_QC_no_ERCC_MT)
test_doublet <- RunPCA(test_doublet)
test_doublet <- RunUMAP(test_doublet, dims = 1:10)

sweep.res.list_test <- paramSweep_v3(test_doublet, PCs = 1:10, sct = T)
sweep.stats_test <- summarizeSweep(sweep.res.list_test, GT = FALSE)
bcmvn_test <- find.pK(sweep.stats_test)

# the optimal pK for our data was manually discerned as maxima in BCmvn distributions
mpK<-as.numeric(as.vector(bcmvn_test$pK[which.max(bcmvn_test$BCmetric)]))
# assumption of 7.5%–10% doublet formation rate per 10,000 cells
nExp_poi <- round(0.075*nrow(test_doublet@meta.data))  
Object_single <- doubletFinder_v3(test_doublet, PCs = 1:10, pN = 0.25, pK = mpK, nExp = nExp_poi, reuse.pANN = FALSE, sct = T)

table(Object_single$DF.classifications_special) # a special parameter like DF.classifications_0.25_0.26_736
dim(Object_single) 

Object_single<-subset(Object_single,subset=DF.classifications_special=="Singlet")
dim(Object_single)
```


Data integration, cell-cycle regression and cell clustering
```{r}
Object.list <- SplitObject(Combine, split.by = "source") 
# Object Combine contains different Object_single mentioned above

# difine cell-cycle genes
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# for mice cell-cycle genes transformed first
human <-useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl")
mouse <- useMart('ensembl',dataset = "mmusculus_gene_ensembl")

s.genes_mice <- getLDS(attributes = c("hgnc_symbol"),
                       filters = "hgnc_symbol", values = s.genes,
                       mart = human,
                       attributesL = c("mgi_symbol"), martL = mouse)

s.genes_mice<-s.genes_mice$MGI.symbol

g2m.genes_mice <- getLDS(attributes = c("hgnc_symbol"),
                         filters = "hgnc_symbol", values = g2m.genes,
                         mart = human,
                         attributesL = c("mgi_symbol"), martL = mouse)

g2m.genes_mice<-g2m.genes_mice$MGI.symbol


mirrow_Object_single <- SCTransform(Object.list$Object_single, verbose = T) 
mirrow_Object_single <- CellCycleScoring(mirrow_Object_single, s.features = s.genes_mice, g2m.features = g2m.genes, set.ident = TRUE)
mirrow_Object_single$CC.Difference <- mirrow_Object_single$S.Score - mirrow_Object_single$G2M.Score
# the difference between the G2M and S phase scores was regressed out of the data
Object.list$Object_single[["CC.Difference"]]<-mirrow_Object_single$CC.Difference
head(Object.list$Object_single[[]])
Object.list$Object_single<- SCTransform(Object.list$Object_single, 
                                vars.to.regress = "CC.Difference",verbose = T) 

features <- SelectIntegrationFeatures(object.list = Object.list, nfeatures = 3000)
Object.list <- PrepSCTIntegration(object.list = Object.list, anchor.features = features)
Object.anchors <- FindIntegrationAnchors(object.list = Object.list, normalization.method = 'SCT', anchor.features = features)
Combined <- IntegrateData(anchorset = Object.anchors, normalization.method = 'SCT')

Combined <- RunPCA(Combined, verbose = FALSE)
Combined <- RunUMAP(Combined, reduction = "pca", dims = 1:30)
Combined <- FindNeighbors(Combined, dims = 1:30)
Combined <- FindClusters(Combined,resolution = 0.5)

# check different resolution
library(clustree)

test <- FindClusters(
  object = Combined,
  resolution = c(seq(0,2,0.5))
)

label_position <- function(labels) {
  if (length(unique(labels)) == 1) {
    position <- as.character(unique(labels))
  } else {
    position <- "mixed"
  }
  return(position)
}

clustree(test@meta.data, prefix = "integrated_snn_res.",
         node_label = "celltype_total",node_label_aggr = "label_position")

test[["res"]]<-test$seurat_clusters
table(test$condition,test$res) # check cell subsets proportional change between different conditions and under different resolutions
```

Cell type identification and annotation
```{r}
# 1) For public datasets, marker genes were identified from the original articles.For information on the methods and data files, please consult the manuscript methods section
DefaultAssay(Combined)<-"SCT"
VlnPlot(object = Combined, features = c("THBD","PECAM1","VWF","CDH5","TEK","ENG")) # EC
VlnPlot(object = Combined, features = c("TNNT2","TTN","MYH6","MYH7"))  # CM
VlnPlot(object = Combined, features = c("FN1","VIM","COL1A1","COL1A2"))  # FB
VlnPlot(object = Combined, features = c("MYH11","ACTA2","CNN1","TAGLN","MYOCD","CALD1","MYLK"))  # SMC
VlnPlot(object = Combined, features = c("CD163","S100A8","CSF1R","C5AR1","CD74"))  # MP

# 2) crosschecked with established markers provided by CellMarker database (available at http://biocc.hrbmu.edu.cn/CellMarker/)
Combined<-FindAllMarkers(Combined,only.pos = T)
top10 <- Atrial_cells_dif %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)

# 3) For more precise atrial tissue specific cell annotation, cell types annotated in the public datasets mentioned above were further used as reference to check the type of our cell populations
dataset1<-Combined
dataset2<-load("~/ScRNA-seq/Nature/Data/Atrial_Total_Cells_Donor.rdata")
test_merge<-merge(dataset1,dataset2)
av <-AverageExpression(test_merge,
                       group.by = "cell_type_for_merge",
                       assays = "SCT")
av=av[[1]] 
head(av) 
cg=names(tail(sort(apply(av, 1, sd)),1000))
View(av[cg,])
View(cor(av[cg,],method = 'spearman'))
Heatmap(cor(av[cg,]),col = viridis(2560))
```

Loading data
```{r}
load("~/ScRNA-seq/Nature/Data/Atrial_Nuclei_Donor.rdata")
load("~/ScRNA-seq/Nature/Data/Atrial_Total_Cells_Donor.rdata")
load("~/ScRNA-seq/Nature/Data/Atrial_CD45_cells_Donor.rdata")
load("~/ScRNA-seq/NCB/Data/Atrial_Cells_Donor&HF.rdata")
load("~/ScRNA-seq/NCB/Data/Atrial_Cells_SR&AF_Right.rdata")
load("~/ScRNA-seq/Own/Data/Atrial_Cells_SR&AF_Left.rdata")
load("~/ScRNA-seq/Own/Data/RFP_cells.rdata")
```

Functional analysis of cell populations
```{r}
# GO analysis
test_markers<-dif_sig_genes # usually use around 200 significantly differential genes
ont<-"BP"
# org.Hs.eg.db for human or org.Mm.eg.db for mouse 
universe <- AnnotationDbi::select(org.Hs.eg.db, keys = rownames(test), columns = "ENTREZID", 
                                  keytype = "SYMBOL", multiVals = "first")$ENTREZID
gene <- AnnotationDbi::select(org.Hs.eg.db, keys = test_markers$gene, 
                              columns = "ENTREZID", keytype = "SYMBOL", multiVals = "first")$ENTREZID
ego <- enrichGO(gene          = gene,
                universe      = universe,
                OrgDb         = org.Hs.eg.db,
                ont           = ont,
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.2)

barplot(ego, showCategory=20)+
  theme(text = element_text(family = "Arial",face = "bold"))+ 
  scale_fill_viridis()

matrix<-ego@result # export ego results for other plots

# GSEA analysis
test_markers<-dif_sig_genes # usually use all significantly differential genes
ont<-"BP"
d <- data.frame(SYMBOL =test_markers$gene)
d$logFC <- test_markers[,"avg_log2FC"]
geneList = d[,2]
names(geneList) = d[,1]
geneList = sort(geneList, decreasing = TRUE)

egse <- gseGO(geneList     = geneList,
              OrgDb        = org.Hs.eg.db,
              ont          = ont,
              keyType      ="SYMBOL",
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = T)

dotplot(egse, showCategory=20) +
  theme(text = element_text(family = "Arial", face = "bold"))+ 
  scale_fill_viridis()

GSEA_result<-egse@result # export GSEA results for other plots
{if (nrow(GSEA_result) <20)
  GSEA_result <- GSEA_result[1:nrow(GSEA_result),] else
    GSEA_result <- GSEA_result[1:20,]}
top10 <- data.frame(GSEA_result$Description,GSEA_result$NES,GSEA_result$p.adjust)
colnames(top10) <- c("Description","NES","p.adjust")
p <- ggplot(data=top10,aes(x=Description,y=NES,fill=p.adjust))+ 
  geom_bar(stat="identity") + coord_flip()+
  theme(panel.background=element_rect(fill='transparent',color='gray'),
        axis.text.y=element_text(color="black",size=12))+ 
  ylim(0,(max(top10$NES)+0.5))+ scale_fill_viridis()+ 
  scale_x_discrete(limits=rev(top10[,1])) +labs(x="",y="Enrichment score")+
  theme(text = element_text(family = "Arial", face = "bold"))
```

Cell–cell communication
```{r}
# NicheNet
ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
save(ligand_target_matrix,file="~/ScRNA-seq/Function_Code/Function_code/NicheNet/ligand_target_matrix.rdata")
ligand_target_matrix[1:5,1:5] # target genes in rows, ligands in columns
lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds"))
save(lr_network,file="~/ScRNA-seq/Function_Code/Function_code/NicheNet/lr_network.rdata")
head(lr_network)
weighted_networks = readRDS(url("https://zenodo.org/record/3260758/files/weighted_networks.rds"))
head(weighted_networks$lr_sig) # interactions and their weights in the ligand-receptor + signaling network
seuratObj<-Combined
Idents(seuratObj)<-seuratObj$celltype_total
seuratObj %>% Idents() %>% table()
table(seuratObj$condition)
nichenet_output = nichenet_seuratobj_aggregate(
  seurat_obj = seuratObj, 
  receiver = "FB", 
  condition_colname = "condition", condition_oi = "one condition", condition_reference = "another condition", 
  sender = "all", 
  ligand_target_matrix = ligand_target_matrix, lr_network = lr_network, weighted_networks = weighted_networks, organism = "human")

# plot for NicheNet
nichenet_output$ligand_activities
nichenet_output$top_ligands
DotPlot(seuratObj, features = nichenet_output$top_ligands %>% rev(), cols = "RdYlBu") + RotatedAxis()
DotPlot(seuratObj, features = nichenet_output$top_ligands %>% rev(), split.by = "condition") + RotatedAxis()
nichenet_output$ligand_target_heatmap +
  theme(text = element_text(family = "Arial", face = "bold",size = 6))
nichenet_output$ligand_activity_target_heatmap

nichenet_output$ligand_receptor_heatmap
nichenet_output$top_receptors
nichenet_output$top_ligands
nichenet_output$top_targets
nichenet_output$ligand_receptor_heatmap_bonafide

# Cellchat
test<-Combined
table(Idents(test))
test[["group"]]<-Idents(test)
cellchat <- createCellChat(test,group.by = "group")
CellChatDB <- CellChatDB.human 
showDatabaseCategory(CellChatDB)
colnames(CellChatDB$interaction)
CellChatDB$interaction[1:4,1:4]
head(CellChatDB$cofactor)
head(CellChatDB$complex)
head(CellChatDB$geneInfo)
CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") 
cellchat@DB <- CellChatDB.use 
cellchat@DB <- CellChatDB
cellchat <- subsetData(cellchat) 
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.human)  

cellchat <- computeCommunProb(cellchat)
cellchat <- filterCommunication(cellchat, min.cells = 10)

cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)

df.net <- subsetCommunication(cellchat) # export data from plots
table(df.net$source,df.net$target) 
EC_FB <- subsetCommunication(cellchat, sources.use = "EC", targets.use = "FB") 

groupSize <- as.numeric(table(cellchat@idents))
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, 
                 weight.scale = T, label.edge= F, title.name = "Number of interactions")+
  theme(text = element_text(family = "Arial", face = "bold"))

# exploring global communication patterns to determine how multiple cell types coordinate with each other through different signaling pathways
selectK(cellchat, pattern = "outgoing")
selectK(cellchat, pattern = "incoming")
nPatterns = n # selected above for outgoing 
cellchat <- identifyCommunicationPatterns(cellchat, 
                                          pattern = "outgoing", 
                                          k = nPatterns,
                                          height=14)
netAnalysis_river(cellchat, pattern = "outgoing")
netAnalysis_dot(cellchat, pattern = "outgoing")
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP")
netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing")

Cell_outgoing<-as.matrix(cellchat@netP$pattern$outgoing$pattern$cell) # export results for plots and statistics
Sig_outgoing<-as.matrix(cellchat@netP$pattern$outgoing$pattern$signaling) # export results for plots and statistics

nPatterns = n # selected above for incoming
cellchat <- identifyCommunicationPatterns(cellchat, 
                                          pattern = "incoming", 
                                          k = nPatterns,
                                          height=14)
netAnalysis_river(cellchat, pattern = "incoming")
netAnalysis_dot(cellchat, pattern = "incoming")
Cell_incoming<-as.matrix(cellchat@netP$pattern$incoming$pattern$cell) # export results for plots and statistics
Sig_incoming<-as.matrix(cellchat@netP$pattern$incoming$pattern$signaling) # export results for plots and statistics

```

Gene score calculation
```{r}
# loading data——Mesenchymal gene scores, EndMT/EMT upregulated gene scores, EndMT/EMT downregulated gene scores, glycolysis gene scores, citric acid cycle (TCA cycle) gene scores, and fatty acid metabolism gene scores 

EMT_genes_mouse<- read.xlsx("~/ScRNA-seq/Data/EMT genes mouse.xlsx", sheet = 1)
EMT_genes_human<- read.xlsx("~/ScRNA-seq/Data/EMT genes human.xlsx", sheet = 1)
TCA_genes_human<- read.xlsx("~/ScRNA-seq/Data/Metabolism.xlsx", sheet = 1)
Glycolysis_genes_human<- read.xlsx("~/ScRNA-seq/Data/Metabolism.xlsx", sheet = 2)
Fatty_acid_metabolism_genes_human <- read.xlsx("~/ScRNA-seq/Data/Metabolism.xlsx", sheet = 3)

# for human
test<-Object
features_Up<-list(EMT_genes_human$`Upregulated.in.EndMT/EMT`)
features_Down<-list(EMT_genes_human$`Downregulated.in.EndMT/EMT`)
features_Mesenchymal<-list(EMT_genes_human$Mesenchymal)

DefaultAssay(test)<-"SCT"
total <- AddModuleScore(
  object = test,
  features = features_Mesenchymal,
  nbin = 24,
  ctrl = 100,
  name = 'EMT_Features'
)

Mesenchymal_counts<-t(data.frame(total$EMT_Features1))

Up <- AddModuleScore(
  object = test,
  features = features_Up,
  nbin = 24,
  ctrl = 100,
  name = 'EMT_up'
)
EMT_up_counts<-t(data.frame(Up$EMT_up1))

Down <- AddModuleScore(
  object = test,
  features = features_Down,
  nbin = 24,
  ctrl = 100,
  name = 'EMT_down'
)

EMT_down_counts<-t(data.frame(Down$EMT_down1))

EMT_counts_combine<-rbind(Mesenchymal_counts,EMT_up_counts)
EMT_counts_combine<-rbind(EMT_counts_combine,EMT_down_counts)
dim(EMT_counts_combine)
rownames(EMT_counts_combine)[1]<-"Mesenchymal Gene Score"
rownames(EMT_counts_combine)[2]<-"EndMT/EMT Upregulated Gene Score"
rownames(EMT_counts_combine)[3]<-"EndMT/EMT Downregulated Gene Score"

EMT_seurat_total<-CreateSeuratObject(EMT_counts_combine,meta.data = test@meta.data)
Idents(EMT_seurat_total)<-EMT_seurat_total$celltype

features<-c("Mesenchymal Gene Score","EndMT/EMT Upregulated Gene Score","EndMT/EMT Downregulated Gene Score")

VlnPlot(EMT_seurat_total,features = features,stack=T,group.by = "celltype",
        flip=T,split.by = "condition",split.plot=T)+
  theme(text = element_text(family = "Arial", face = "bold")) + scale_fill_brewer(palette = "Pastel2")

RidgePlot(EMT_seurat_total,features = features)+ scale_fill_viridis(discrete=T)

# for mouse
test<-Object
features_Up<-list(EMT_genes_mouse$`Upregulated.in.EndMT/EMT`)
features_Down<-list(EMT_genes_mouse$`Downregulated.in.EndMT/EMT`)
features_Mesenchymal<-list(EMT_genes_mouse$Mesenchymal)

human <-useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl")
mouse <- useMart('ensembl',dataset = "mmusculus_gene_ensembl")

TCA_genes_mice <- getLDS(attributes = c("hgnc_symbol"),
                       filters = "hgnc_symbol", values = TCA_genes_human$Symbol,
                       mart = human,
                       attributesL = c("mgi_symbol"), martL = mouse)

TCA_genes_mice <-TCA_genes_mice$MGI.symbol

Glycolysis_genes_mice <- getLDS(attributes = c("hgnc_symbol"),
                         filters = "hgnc_symbol", values = Glycolysis_genes_human$Symbol,
                         mart = human,
                         attributesL = c("mgi_symbol"), martL = mouse)

Glycolysis_genes_mice<-Glycolysis_genes_mice$MGI.symbol

Fatty_acid_metabolism_genes_mice <- getLDS(attributes = c("hgnc_symbol"),
                         filters = "hgnc_symbol", values = Fatty_acid_metabolism_genes_human$Symbol,
                         mart = human,
                         attributesL = c("mgi_symbol"), martL = mouse)

Fatty_acid_metabolism_genes_mice<-Fatty_acid_metabolism_genes_mice$MGI.symbol

TCA_genes_mice<-list(TCA_genes_mice)
Glycolysis_genes_mice<-list(Glycolysis_genes_mice)
Fatty_acid_metabolism_genes_mice<-list(Fatty_acid_metabolism_genes_mice)

DefaultAssay(test)<-"SCT"
TCA <- AddModuleScore(
  object = test,
  features = TCA_genes_mice,
  nbin = 24,
  ctrl = 100,
  name = 'TCA'
)

TCA_counts<-t(data.frame(TCA$TCA1))

Glycolysis <- AddModuleScore(
  object = test,
  features = Glycolysis_genes_mice,
  nbin = 24,
  ctrl = 100,
  name = 'Glycolysis'
)
Glycolysis_counts<-t(data.frame(Glycolysis$Glycolysis1))

Fatty_acid_metabolism <- AddModuleScore(
  object = test,
  features = Fatty_acid_metabolism_genes_mice,
  nbin = 24,
  ctrl = 100,
  name = 'Fatty_acid_metabolism'
)

Fatty_acid_metabolism_counts<-t(data.frame(Fatty_acid_metabolism$Fatty_acid_metabolism1))
Metabolism_counts_combine<-rbind(Glycolysis_counts,TCA_counts)
Metabolism_counts_combine<-rbind(Metabolism_counts_combine,Fatty_acid_metabolism_counts)
dim(Metabolism_counts_combine)
rownames(Metabolism_counts_combine)[1]<-"Glycolysis Gene Score"
rownames(Metabolism_counts_combine)[2]<-"TCA Gene Score"
rownames(Metabolism_counts_combine)[3]<-"Fatty_acid_metabolism Gene Score"

Metabolism_seurat_total<-CreateSeuratObject(Metabolism_counts_combine,meta.data = test@meta.data)
rownames(Metabolism_seurat_total)

features<-c("Glycolysis Gene Score","TCA Gene Score","Fatty-acid-metabolism Gene Score")

VlnPlot(Metabolism_seurat_total,features = features,stack=T,group.by = "celltype",
        flip=T,split.by = "condition",split.plot=T)+
  theme(text = element_text(family = "Arial", face = "bold")) + scale_fill_brewer(palette = "Pastel2")
```

Cell trajectory analysis
```{r}
test_trajectory<-Object # Given an object to cell trajectory analysis
table(Idents(test_trajectory),test_trajectory$condition) # Check the condition groups of Object
Idents(test_trajectory)<-test_trajectory$condition 
markers <- FindMarkers(test_trajectory, verbose=T,ident.1 = "condition_1",ident.2 = "condition_2")  # find differential genes between different conditions
ordering_genes <- rownames(unique(subset(markers, p_val_adj < 0.05 & abs(avg_log2FC)>0.5))) 
{
  test_trajectory_matrix<- test_trajectory@assays$SCT@counts
  sample_ann<-test_trajectory@meta.data
  feature_ann<-data.frame(gene_id=rownames(test_trajectory_matrix),gene_short_name=rownames(test_trajectory_matrix))
  rownames(feature_ann)<-rownames(test_trajectory_matrix)
  pd <- new("AnnotatedDataFrame", data = sample_ann)
  fd <- new("AnnotatedDataFrame", data = feature_ann)
  test_trajectory_CDS <- newCellDataSet(test_trajectory_matrix,phenoData = pd, 
                           featureData = fd,expressionFamily=negbinomial.size())
}  

test_trajectory_CDS <- estimateSizeFactors(test_trajectory_CDS)
test_trajectory_CDS <- estimateDispersions(test_trajectory_CDS)
test_trajectory_CDS <- detectGenes(test_trajectory_CDS, min_expr = 0.1)
test_trajectory_CDS <- setOrderingFilter(test_trajectory_CDS, ordering_genes)
plot_ordering_genes(test_trajectory_CDS)
test_trajectory_CDS <- reduceDimension(test_trajectory_CDS, max_components = 2,method = 'DDRTree')
test_trajectory_CDS  <- orderCells(test_trajectory_CDS)

GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$condition)[,"condition_sepcific"]
    return(as.numeric(names(T0_counts)[which
                                       (T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}
# call orderCells again using the root_state argument to specify the beginning
test_trajectory_CDS <- orderCells(test_trajectory_CDS, root_state = GM_state(test_trajectory_CDS))
test_trajectory_CDS_after_root<-test_trajectory_CDS

plot_cell_trajectory(test_trajectory_CDS, color_by = "State")+
  theme(text = element_text(family = "Arial", size = 16, face = "bold"))
plot_cell_trajectory(test_trajectory_CDS, color_by = "Pseudotime")+
  theme(text = element_text(family = "Arial", size = 16, face = "bold"))
# export data for statistics
colnames(pData(test_trajectory_CDS))
table(pData(test_trajectory_CDS)$celltype)
table(pData(test_trajectory_CDS)$State,pData(test_trajectory_CDS)$celltype)
table(pData(test_trajectory_CDS)$State,pData(test_trajectory_CDS)$condition)

# plot genes in pseudotime
my_genes <- row.names(subset(fData(test_trajectory_CDS),gene_short_name %in% my_genes))
cds_subset <- test_trajectory_CDS[my_genes,]
plot_genes_in_pseudotime(cds_subset, color_by = "condition",ncol=3,cell_size=0.1)+
  theme(text = element_text(family = "Arial", size = 16, face = "bold"))
plot_genes_in_pseudotime(cds_subset, color_by = "Pseudotime",ncol=4,cell_size=0.5)

plot_genes_branched_pseudotime(cds_subset,
                               branch_point = 1,
                               color_by = "condition",
                               ncol = 2, branch_labels=c("cell fate 1","cell fate 2"))+
  theme(text = element_text(family = "Arial", size = 16, face = "bold"))

# calculate p-values along the pseudotime
diff_test_res <- differentialGeneTest(test_trajectory_CDS,fullModelFormulaStr = "~sm.ns(Pseudotime)")
diff_test_res <- differentialGeneTest(cds_subset,fullModelFormulaStr = "~sm.ns(Pseudotime)")

res<-diff_test_res[,c("gene_short_name", "pval", "qval")]
genes<-subset(res,qval<0.01)

# heatmap showing significant genes
sig_gene_names <- row.names(subset(diff_test_res, qval < 0.01))
plot_pseudotime_heatmap(test_trajectory_CDS[sig_gene_names,],
                        num_clusters = 3,
                        cores = 1,
                        show_rownames = T)

# BEAM test branch-dependent genes
BEAM_res <- BEAM(test_trajectory_CDS, branch_point = 1, cores = 1)
BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]

plot_genes_branched_heatmap(test_trajectory_CDS[row.names(subset(BEAM_res,qval < 1e-4)),],
                            branch_point = 1,
                            num_clusters = 4,
                            cores = 1,
                            use_gene_short_name = T,
                            show_rownames = T)

# Integrating state into Seurat object
test_trajectory[["State"]]<-test_trajectory_CDS$State
table(test_trajectory$State)
table(test_trajectory_CDS$State)

glycolysis<-c("Aldoa","Bpgm","Eno1","Eno2","Gapdh","Gpi","Hk1","Hk2","Hkdc1","Pfkl","Pfkm","Pgam1",
              "Pgam2","Pgam4","Pgk1","Pklr","Pkm","Tpi1")

TCA<-c("Aco1","Aco2","Cs","Dhtkd1","Dlat","Dld","Dlst","Fh","Idh1","Idh2","Idh3a","Idh3b","Idh3g","Mdh1",
       "Mdh1b","Mdh2","Nnt","Ogdh","Ogdhl","Pdha1","Pdha2","Pdhb","Sdha","Sdhb","Sdhc","Sdhd",
       "Sucla2","Suclg1","Suclg2")

FA<-c("Abcd1","Abcd2","Abcd3","Acaa1","Acad11","Acadl","Acadm","Acads","Acadvl","Acot7","Acot8",
      "Acox1","Acox2","Acox3","Acoxl","Adipoq","Amacr","Bdh2","Cpt1a","Cpt1b","Hsd17b4","Irg1","Lep","Lipe",
      "Lpin1","Lpin2","Lpin3","Mcee","Mmaa","Mut","Pcca","Pccb","Pex13","Pex2","Pex7","Phyh","Pla2g15",
      "Ppard","Scp2","Sesn2","Slc25a17","Slc27a2","Slc27a4","Cpt1c","Cpt2","Crat","Crot","Cyp4a11","Cyp4f12","Cyp4f2","Cyp4f3","Decr1","Ech1","Echs1",
      "Eci1","Ehhadh","Etfdh","Faah","Hacl1","Hadh","Hadha","Hadhb","Hao1")

DotPlot(test_trajectory,features = glycolysis, group.by = "State",split.by = "condition")
DotPlot(test_trajectory,features = TCA, group.by = "State",split.by = "condition")
DotPlot(test_trajectory,features = FA, group.by = "State",split.by = "condition")

Idents(test_trajectory)<-test_trajectory$State
# major metabolic genes, especially those involved in fibrosis, were analyzed
features<-c("Hk1","Hk2","Pfkfb3","Pkm","Pdk1","Pdk2","Pdk3","Pdk4","Cpt1a","Cpt1b","Acaca","Acacb")
DotPlot(test_trajectory,features = features)+ scale_color_distiller(palette = "Spectral")+
  coord_flip()+
  theme(text = element_text(family = "Arial", size = 16, face = "bold"))

# calculate metabolic gene scores within different cell states
TCA_genes_mice<-list(TCA_genes_mice)
Glycolysis_genes_mice<-list(Glycolysis_genes_mice)
Fatty_acid_metabolism_genes_mice<-list(Fatty_acid_metabolism_genes_mice)
# AddModuleScore used to calculate metabolic gene scores as mentioned above, Metabolism_counts_combine was obtained
# rownames(Metabolism_counts_combine) —— "Glycolysis Gene Score", "TCA Gene Score", "Fatty_acid_metabolism Gene Score"
Metabolism_seurat_total<-CreateSeuratObject(Metabolism_counts_combine,meta.data = test@meta.data)
state_n<-subset(Metabolism_seurat_total,subset=State=="n")
Metabolism_state_n_matrix<-t(data.frame(state_n@assays$RNA@counts))
# export metabolic gene score of each cell for statistics
write.table(Metabolism_state_n_matrix, file = "~/ScRNA-seq/Data/Metabolism_state_n_matrix.csv", 
            row.names = TRUE, sep = ",", col.names = TRUE) 
```


































